### PR TITLE
fixed negative  children_count after deleting categories

### LINKED ISSUE #27969: children_count can become negative after deleting categories
### Preconditions

1. Install clean Magento 2.3.4

### Steps to reproduce

1. Log in into admin backend
2. Navigate to Catalog -> Categories
3. Click "Add Subcategory"
4. Fill category name with "Parent 1", Save
5. Click "Add Subcategory"
6. Fill category name with "Child 1", Save
7. Click "Add Subcategory"
8. Fill category name with "Child 2", Save
9. Click "Add Subcategory"
10. Fill category name with "Child 3", Save. At this point this category should have level = 5

11. Click on "Child 1" category
12. Delete category

### Expected result

1. Categories are deleted and children_count for remaining categories are valid eg. catalog_category_entity look as follows:

```
1    3    0    2020-02-05 18:38:54    2020-04-24 10:50:38    1    0    0    2
2    3    1    2020-02-05 18:38:54    2020-04-24 10:50:38    1/2    1    1    1
15    3    2    2020-04-24 10:49:35    2020-04-24 10:50:38    1/2/15    1    2    0
```

### Actual result

1. Categories are deleted but children_count for remaining categories are negative -  catalog_category_entity look as follows:

```
1    3    0    2020-02-05 18:38:54    2020-04-24 10:50:38    1    0    0    -1
2    3    1    2020-02-05 18:38:54    2020-04-24 10:50:38    1/2    1    1    -2
15    3    2    2020-04-24 10:49:35    2020-04-24 10:50:38    1/2/15    1    2    -3
```

screenshot

### PR DESCRIPTION
**Pull request by ProkopovVitaliy**
### Fixed Issues (if relevant)

1. magento/magento2#27969: children_count can become negative after deleting categories

### Manual testing scenarios (*)

1.Log in into admin backend
2.Navigate to Catalog -> Categories
3.Click "Add Subcategory"
4.Fill category name with "Parent 1", Save
5.Click "Add Subcategory"
6.Fill category name with "Child 1", Save
7.Click "Add Subcategory"
8.Fill category name with "Child 2", Save
9.Click "Add Subcategory"
10.Fill category name with "Child 3", Save. At this point this category should have level = 5
11.Click on "Child 1" category
12.Delete category

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Echo** [comment]
I will take care of test coverage.

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-7513](https://jira.corp.magento.com/browse/ENGCOM-7513) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1.Log in into admin backend
2.Navigate to Catalog -> Categories
3.Click "Add Subcategory"
4.Fill category name with "Parent 1", Save
5.Click "Add Subcategory"
6.Fill category name with "Child 1", Save
7.Click "Add Subcategory"
8.Fill category name with "Child 2", Save
9.Click "Add Subcategory"
10.Fill category name with "Child 3", Save. At this point this category should have level = 5
11.Click on "Child 1" category
12.Delete category
13. See `catalog_category_entity` table

**Before:**  :heavy_multiplication_x: `children_count` for remaining categories have negative meanings
screenshotqwerty

**After:** :heavy_check_mark: children_count for remaining categories are valid 
screenshotqwerty2

Also tested with few roots categories and Childs.

**Comment by ProkopovVitaliy** [comment]
@magento run all tests

**Comment by ProkopovVitaliy** [comment]
@magento run all tests

**Comment by ProkopovVitaliy** [comment]
@VladimirZaets I think failing tests are not related to the changes presented in the current PR 🤔.

**Comment by VladimirZaets** [comment]
@ProkopovVitaliy I reran tests to be sure, thanks

**Comment by ProkopovVitaliy** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by ProkopovVitaliy** [comment]
@magento run Unit Tests

**Comment by ProkopovVitaliy** [comment]
@magento run Static Tests

**Comment by ProkopovVitaliy** [comment]
@magento run Unit Tests

**Comment by ProkopovVitaliy** [comment]
@magento run Static Tests

**Comment by ProkopovVitaliy** [comment]
@magento run Unit Tests

File: app/code/Magento/Catalog/Test/Unit/Model/ResourceModel/Category/AggregateCountTest.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * @var AdapterInterface|MockObject
+     */
+    protected $connectionMock;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setUp()
```
Comment (dmytro-ch) [inline app/code/Magento/Catalog/Test/Unit/Model/ResourceModel/Category/AggregateCountTest.php]: Hello @ProkopovVitaliy,
could you please add a return type definition here for compatibility with PHPUnit 8?
```suggestion
    public function setUp(): void
```

**Review by dmytro-ch — CHANGES_REQUESTED** [review]

**Comment by ProkopovVitaliy** [comment]
@magento run all tests

**Review by dmytro-ch — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @dmytro-ch, thank you for the review.
[ENGCOM-7513](https://jira.corp.magento.com/browse/ENGCOM-7513) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
**Notice:** QA not applicable
New commits are connected to tests

**Pull request merged**